### ISSUE TITLE
`unoconv` listener process may become a zombie

### ISSUE THREAD

**Issue by gulien** [gotenberg/gotenberg#406]
On the Live Demo, the `unoconv` listener process did become a zombie for some reason. 

While `unoconv` is still working (it creates a dedicated LibreOffice instance in such a scenario), the conversions are way slower.

In order to prevent this from happening, we have to:

1. Check the state of the `unoconv` listener process.
2. If state is invalid (no such process / zombie), try to restart the listener.
3. If restart fails, `/health` endpoint should returns an error.

**Comment by omni-htg** [comment]
Hey @gulien !
I've noticed that unoconv's successor, unoserver, changed the way it creates the listener so it can be "daemonized" -- allowing the main python process to exit.
I'm not sure if this fixes the memory leaks experienced (as I still haven't tested whether the guilt resides on unoconv or libreoffice), but IMO this means that the proposed solution might not cover the whole picture.

Will report back if I find anything else on the matter!

PS. If certain UNO interactions can be retrofitted from unoserver to unoconv, I believe that LibreOffice's PDF/UA conversion might be possible on Gotenberg.

**Comment by gulien** [comment]
Actually ,`unoserver` won't change anything, because it works like `unoconv --listener` 😄 i.e., starting a LibreOffice instance that accepts connection on a port.

I'm currently working on my "own" implementation of the listener, that may help reduce drastically the memory leaks (at least over time) while resolving also the "zombie" issue.

And, of course, with more PDF formats available 😄 (namely, `PDF/A-2b`, `PDF/A-3b`).

**Comment by omni-htg** [comment]
Haha, yeah, that part is definitely close to unoconv -- but what I meant was the possibility to daemon the listener.
In that way, you wouldn't be able to check if it's zombiefied or not, as only the "main" LibreOffice process (and it's children) are left, right?

> I'm currently working on my "own" implementation of the listener

Can't wait!
I'm also working on improving unoserver in ways that might solve this problem and [hopefully can benefit gotenberg later](https://wiki.openoffice.org/wiki/API/Tutorials/PDF_export#PDF_Export_Configuration).

>  with more PDF formats available

Dang, were you able to do it with `SelectPdfVersion`?

**Comment by gulien** [comment]
> In that way, you wouldn't be able to check if it's zombiefied or not, as only the "main" LibreOffice process (and it's children) are left, right?

The zombie issue happens because I forgot to wait for the unoconv listener process in the parent process (i.e., Gotenberg) 😅 

My listener will launch directly a LibreOffice process, so Gotenberg will somehow work like `unoserver`. The main advantage for me is that I can keep track of the LibreOffice's state.

> Dang, were you able to do it with SelectPdfVersion?

Yes, it's really stupid in fact 😄:

* `SelectPdfVersion = 1` => `PDF/A-1a`
* `SelectPdfVersion = 2` => `PDF/A-2b`
* `SelectPdfVersion = 3` => `PDF/A-3b`

**Comment by omni-htg** [comment]
> The zombie issue happens because I forgot to wait for the unoconv listener process in the parent process (i.e., Gotenberg) 😅

Haha it happens!
It can also take a while for the whole listener to go from startup to ready (at least in my experience with `unoserver`)...

> The main advantage for me is that I can keep track of the LibreOffice's state.

What kind of "state information" are we talking about here? Like memory usage, or is there another metric that might be useful?

> Yes, it's really stupid in fact 😄

I see, that's pretty straightforward 😄 
For some reason I couldn't find those options in the docs I've read, so thanks!
This should also mean that you can include `PDFUACompliance` and other options, right? 👍 

In regards to memory leaks:  
In the few tests I've performed with basic bash loops, I've also experienced the LOo process tree growing in MEM without really freeing it up after each conversion -- though not as extreme as seen in #407.

**Comment by gulien** [comment]
> It can also take a while for the whole listener to go from startup to ready.

Yep 😄 first, there is the first start (exit code `81`), then you have to restart LibreOffice and wait for its socket to be ready.

> What kind of "state information" are we talking about here?

State = is it running or not :trollface: 

> For some reason I couldn't find those options in the docs I've read, so thanks!
This should also mean that you can include PDFUACompliance and other options, right? 👍

I had to check the LibreOffice issue tracker lol
Not sure about `PDFUACompliance`, all I know is that the different PDF formats are recognized in most if not all PDF online validators (i.e., they recognize that a PDF converted with Gotenberg is indeed `PDF/A-3b` for instance).

> I've also experienced the LOo process tree growing in MEM without really freeing it up after each conversion.

Dunno why it happens even when `unoconv` (and LibreOffice) are killed. AFAIK, it should free the memory almost immediatly. If you have/find anything about this issue, do not hesitate to share it with me 👍 

> though not as extreme as seen in #407.

Yep, but @jesseinit is converting large documents, so I guess, the larger the documents, the larger the memory leak.

**Comment by omni-htg** [comment]
> AFAIK, it should free the memory almost immediatly.

That is my experience as well. Maybe I should try with larger documents later on to see if that has an effect.

> I had to check the LibreOffice issue tracker lol

Nice, thank you for that, I'll check it out.
I'd still recommend you include the option for PDF/UA output in some way at some point (especially when it's as "easy" as including the `PDFUACompliance` option).

> State = is it running or not

Heh.
What gives you back the exit code 81? `unoconv`?
The sockets check is pretty clever, tbh -- I was checking for the processes' state.

**Comment by gulien** [comment]
> I'd still recommend you include the option for PDF/UA output in some way at some point (especially when it's as "easy" as including the PDFUACompliance option).

Not sure how to do that with `unoconv` 🤔 

> What gives you back the exit code 81? unoconv?

LibreOffice actually; `unoconv` is just for interacting with the LibreOffice instance (there is no UNO library in GO alas).

**Comment by omni-htg** [comment]
> that's a TCP socket btw.

Thank you! Fortunately I've been messing with them trying to understand how to perform a "is there a file processing" check (based on [turulomio's unogenerator](https://github.com/turulomio/unogenerator/blob/e4d9979b138f9bc774fe4432649d03d2b805eb35/unogenerator/server.py), so it shouldn't be too hard to apply that on a "am I ready" one (I hope...).

> there is no UNO library in Go alas

*Shares pain in C#*

> Not sure how to do that with unoconv 🤔

It might not be _as straightforward_ as other options, but if what I could do in unoserver can be applied here, you should be able to take advantage of OOo's `FilterData` object to send whatever (PDF) export option/s you want as an array of `com.sun.star.beans.PropertyValue`.

**Comment by gulien** [comment]
>  I've been messing with them trying to understand how to perform a "is there a file processing" check

My solution is to work with a "lock mechanism", i.e., if there is a lock, there is a file being processed by LibreOffice. Not sure how it would translate in C#, but it works fine in Go 👍 

> you should be able to take advantage of OOo's FilterData object to send whatever (PDF) export option/s you want as an array of com.sun.star.beans.PropertyValue.

Is that currently feasible with `unoserver`? Or it's just your custom variant? Anyway, I'm really interested in your solution 😄

**Comment by omni-htg** [comment]
>  work with a "lock mechanism", i.e., if there is a lock, there is a file being processed by LibreOffice.

Clever. Will def check that out and see if I can observe the same behaviour 😄 
Using Python for all of this makes it a bit easier to have access to tools.
You can also check the connections established to the main `soffice` process.

> Is that currently feasible with `unoserver`?

Yes, it is -- as you said, they are just a way to communicate with UNO.
In [unoserver's converter.py you have these properties](https://github.com/unoconv/unoserver/blob/ce75e6bbe1afea67e2bb4ee421d503039e4c3ab4/src/unoserver/converter.py#L191-L194) that you "send" to the listener -- by adding FilterData as another one you are able to do this.
Just append what you need inside of that FilterData property, ie.:
```python
output_props = (
    PropertyValue(Name="FilterName", Value=filtername),
    PropertyValue(Name="Overwrite", Value=True),
    PropertyValue(Name="FilterData", Value=uno.Any("[]com.sun.star.beans.PropertyValue", (PropertyValue(Name="PDFUACompliance", Value=True), )))
)
```

This is not verbatim, so give it a spin.

**Comment by gulien** [comment]
I might have to write my own Python script in the future 😅

**Comment by omni-htg** [comment]
> I might have to write my own Python script in the future 😅

I don't want to promise anything since it might take a while -- but once I can get this sorted out on my side, I can try and help more with a repo or pull req.

**Comment by gulien** [comment]
That would be great 🚀

**Comment by omni-htg** [comment]
Oh, btw, maybe it was because I am so dumb with python, but I just found out that +`unoserver` _does not_ properly clean up the tmp folder it generates for libreoffice, even after a controlled SIGTERM.
It does successfully exit all the loo processes, though.
